File: frameworks/python/fastapi/Implementation/routes/sse.py
from collections.abc import AsyncIterator

from fastapi import APIRouter
from fastapi.sse import EventSourceResponse

from payloads import Item, Payloads


def router(p: Payloads) -> APIRouter:
    """sse: items.medium's rows as server-sent events, through FastAPI's own support for them. Each
    row the handler yields is serialised by Pydantic, from the declared item type, as one event."""
    routes = APIRouter()

    @routes.get("/sse/medium", response_class=EventSourceResponse)
    async def medium() -> AsyncIterator[Item]:
        for row in p.medium.items:
            yield row

    return routes

File: frameworks/rust/loco/Implementation/controllers/sse.rs
use axum::response::sse::{Event, Sse};
use futures_util::stream::{self, StreamExt};
use loco_rs::prelude::*;

use crate::Payloads;

// rb:handler sse.medium
/// sse: items.medium's rows as server-sent events, through axum's own response for them, which
/// writes each row serde_json serialises as the data of one event. Loco has none of its own.
async fn medium(SharedStore(p): SharedStore<&'static Payloads>) -> Result<Response> {
    Ok(Sse::new(stream::iter(&p.medium.items).map(|row| Event::default().json_data(row))).into_response())
}

pub fn routes() -> Routes {
    Routes::new().prefix("sse").add("/medium", get(medium))
}
